src/circulating_supply/circulating_supply_provider.rs

```diff
@@ -0,0 +1,150 @@
+    get_account_state, get_code_version, is_bug_inside, TRANSFERS_ENABLED,
+};
+use crate::circulating_supply::user_balance::get_user_balance;
+use crate::db_adapters::accounts::collect_lockups_for_block;
+use crate::db_adapters::blocks::closest_block_for;
+use crate::db_adapters::circulating_supply::{
+    add_circulating_supply, get_precomputed_circulating_supply,
+};
+use crate::models;
+use crate::models::circulating_supply::CirculatingSupply;
+use crate::models::Block;
+use actix::Addr;
+use actix_web::rt::time;
+use std::ops::Div;
+use std::str::FromStr;
+use std::time::SystemTime;
+use tracing::info;
+
+const DAY: std::time::Duration = std::time::Duration::from_secs(86400);
+
+pub(crate) async fn compute_circulating_supply(view_client: Addr<near_client::ViewClientActor>) {
+    let mut current_day = TRANSFERS_ENABLED;
+    while current_day < get_now_timestamp() {
+        compute_circulating_supply_for(&view_client, current_day).await;
+        current_day += DAY.as_nanos() as u64;
+    }
+
+    // We ping the number twice a day to be sure that we computed it, and to deliver it earlier.
+    // If the value is already in the DB, we will just return almost immediately
+    let mut interval = time::interval(DAY.div(2));
+    loop {
+        compute_circulating_supply_for(&view_client, get_now_timestamp()).await;
+        interval.tick().await;
+    }
+}
+
+fn get_now_timestamp() -> u64 {
+    SystemTime::now()
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .expect("Time went backwards")
+        .as_nanos() as u64
+}
+
+async fn compute_circulating_supply_for(
+    view_client: &Addr<near_client::ViewClientActor>,
+    request_timestamp: u64,
+) {
+    let pool = models::establish_connection();
+
+    let block = find_last_yesterday_block_for(request_timestamp, &pool).await;
+    let block_timestamp = block
+        .block_timestamp
+        .to_u64()
+        .expect("`block_timestamp` expected to be u64");
+    let block_height = block
+        .block_height
+        .to_u64()
+        .expect("`block_height` expected to be u64");
+    let total_supply = u128::from_str_radix(&*block.total_supply.to_string(), 10)
+        .expect("`total_supply` expected to be u128");
+
+    match get_precomputed_circulating_supply(block_timestamp, &pool).await {
+        Ok(None) => {
+            info!(
+                target: crate::INDEXER_FOR_EXPLORER,
+                "Computing circulating supply for the timestamp {}", block_timestamp
+            );
+        }
+        Ok(Some(supply)) => {
+            info!(
+                target: crate::INDEXER_FOR_EXPLORER,
+                "Circulating supply for the timestamp {} was already computed: {}",
+                block_timestamp,
+                supply
+            );
+            return;
+        }
+        Err(err) => {
+            panic!("Error {}", err);
+        }
+    }
+
+    let lockups = collect_lockups_for_block(block_height, &pool).await;
```

Comment:
Are they list of lockup account ids? The naming is misleading since I read it as such that the method already returns a list of lockup information structs with all the necessary information about them.